Lib/copy.py

```diff
@@ -126,16 +126,20 @@ def deepcopy(x, memo=None, _nil=[]):
     if cls in _atomic_types:
         return x
 
-    d = id(x)
     if memo is None:
+        if cls in _builtin_iterables and not x:
+            return cls()
+        d = id(x)
         memo = {}
     else:
+        d = id(x)
         y = memo.get(d, _nil)
         if y is not _nil:
             return y
 
-    copier = _deepcopy_dispatch.get(cls)
-    if copier is not None:
+    if cls in _builtin_iterables and not x:
+        y = cls()
+    elif copier := _deepcopy_dispatch.get(cls):
```

Comment:
Reverting these lines would remove the benefits of the optimisation for empty iterables nested inside other containers something e.g.`[[], (), {}, set(), frozenset()]` but in turn it seems like it would remove the 3% performance regression that this PR would otherwise introduce to the common case.
That actually seems like it might be a better tradeoff, let me know what you think. I'm happy to run some more benchmarks then to verify that we are not introducing any performance regression on the common case.

This is what the numbers would look like `opt-deep` (this PR), `opt-empty` (this PR without these lines):
```
+------------------------------------+----------+------------------------+------------------------+
| Benchmark                          | baseline | opt-deep               | opt-empty              |
+====================================+==========+========================+========================+
| deepcopy dict                      | 1.88 us  | 1.93 us: 1.03x slower  | not significant        |
+------------------------------------+----------+------------------------+------------------------+
| deepcopy empty tuple               | 287 ns   | 49.4 ns: 5.81x faster  | 49.8 ns: 5.77x faster  |
+------------------------------------+----------+------------------------+------------------------+
| deepcopy empty frozenset           | 1.46 us  | 67.4 ns: 21.63x faster | 67.1 ns: 21.70x faster |
+------------------------------------+----------+------------------------+------------------------+
| deepcopy empty list                | 327 ns   | 64.4 ns: 5.08x faster  | 67.1 ns: 4.88x faster  |
+------------------------------------+----------+------------------------+------------------------+
| deepcopy empty set                 | 1.44 us  | 70.1 ns: 20.60x faster | 67.9 ns: 21.26x faster |
+------------------------------------+----------+------------------------+------------------------+
| deepcopy empty dict                | 329 ns   | 69.1 ns: 4.76x faster  | 68.1 ns: 4.83x faster  |
+------------------------------------+----------+------------------------+------------------------+
| deepcopy multiple empty containers | 4.16 us  | 1.27 us: 3.28x faster  | not significant        |
+------------------------------------+----------+------------------------+------------------------+
| Geometric mean                     | (ref)    | 5.72x faster           | 4.85x faster           |
+------------------------------------+----------+------------------------+------------------------+
```

Reply:
I decided to revert these lines in d87901edccdb811075711f7a6d1651044afc8086 so now this change doesn't seem to influence the performance of the common case anymore.